src/Features/CSharp/Portable/SignatureHelp/AbstractCSharpSignatureHelpProvider.cs

```diff
@@ -68,5 +96,293 @@ protected static SymbolDisplayPart NewLine()
         protected IList<TaggedText> GetAwaitableUsage(IMethodSymbol method, SemanticModel semanticModel, int position)
 #pragma warning restore CA1822 // Mark members as static
             => SpecializedCollections.EmptyList<TaggedText>();
+
+        // TODO2 move to a separate file before merging
+        internal static class LightweightOverloadResolution
+        {
+            internal static void RefineOverloadAndPickParameter(Document document, int position, SemanticModel semanticModel,
+                ImmutableArray<IMethodSymbol> candidates, SeparatedSyntaxList<ArgumentSyntax> arguments,
+                out IMethodSymbol? currentSymbol, out int parameterIndex)
+            {
+                var semanticFactsService = document.GetRequiredLanguageService<ISemanticFactsService>();
+                if (candidates.Length == 1)
+                {
+                    // The compiler told us the correct overload or we only have one choice, but we need to find out the parameter to highlight given cursor position
+                    currentSymbol = candidates[0];
+                    _ = FindParameterIndexIfCompatibleMethod(arguments, currentSymbol, position, semanticModel, semanticFactsService, out parameterIndex);
+                }
+                else
+                {
+                    (currentSymbol, parameterIndex) = GuessCurrentSymbolAndParameter(arguments, candidates, position, semanticModel, semanticFactsService);
+                }
+            }
+
+            /// <summary>
+            /// If the symbol could not be bound, we could be dealing with a partial invocation, we'll try to find a possible overload.
+            /// </summary>
+            private static (IMethodSymbol? symbol, int parameterIndex) GuessCurrentSymbolAndParameter(
+                SeparatedSyntaxList<ArgumentSyntax> arguments, ImmutableArray<IMethodSymbol> methodGroup, int position,
+                SemanticModel semanticModel, ISemanticFactsService semanticFactsService)
+            {
+                if (arguments.Count != 0)
+                {
+                    foreach (var method in methodGroup)
+                    {
+                        if (FindParameterIndexIfCompatibleMethod(arguments, method, position, semanticModel, semanticFactsService, out var parameterIndex))
+                        {
+                            return (method, parameterIndex);
+                        }
+                    }
+                }
+
+                // Note: Providing no recommendation if no arguments allows the model to keep the last implicit choice
+                return (null, -1);
```

Comment:
would `return default;` work here?  if the caller sees that the symbol returned is null, do they even bother with the parameterIdnex?

Reply:
It might work, but I think I prefer explicit tuple values.

Reply:
wfm.